esp-hal/src/soc/esp32s3/mmu.rs

```diff
@@ -0,0 +1,166 @@
+const ENTRY_VALID_VAL_MASK: u32 = 0x3fff;
+const ICACHE_MMU_SIZE: usize = 0x800;
+
+pub(super) const ENTRY_ACCESS_SPIRAM: u32 = 1 << 15;
+pub(super) const PAGE_SIZE: usize = 0x10000;
+pub(super) const TABLE_SIZE: usize = ICACHE_MMU_SIZE / core::mem::size_of::<u32>();
+
+extern "C" {
+    /// Set DCache mmu mapping.
+    ///
+    /// [`ext_ram`]: u32 ENTRY_ACCESS_FLASH for flash, ENTRY_ACCESS_SPIRAM for spiram, ENTRY_INVALID for invalid.
+    /// [`vaddr`]: u32 Virtual address in CPU address space.
+    /// [`paddr`]: u32 Physical address in external memory. Should be aligned by psize.
+    /// [`psize`]: u32 Page size of DCache, in kilobytes. Should be 64 here.
+    /// [`num`]: u32 Pages to be set.
+    /// [`fixes`]: u32 0 for physical pages grow with virtual pages, other for virtual pages map to same physical page.
+    pub(super) fn cache_dbus_mmu_set(
+        ext_ram: u32,
+        vaddr: u32,
+        paddr: u32,
+        psize: u32,
+        num: u32,
+        fixed: u32,
+    ) -> i32;
+
+    fn Cache_Invalidate_Addr(addr: u32, size: u32);
+    fn Cache_WriteBack_All();
+    fn rom_Cache_WriteBack_Addr(addr: u32, size: u32);
+}
+
+#[procmacros::ram]
+pub(super) fn last_mapped_index() -> Option<usize> {
+    let mmu_table_ptr = DR_REG_MMU_TABLE as *const u32;
+    (0..TABLE_SIZE)
+        .rev()
+        .find(|&i| unsafe { mmu_table_ptr.add(i).read_volatile() } != ENTRY_INVALID)
+}
+
+#[procmacros::ram]
+pub(super) fn index_to_data_address(index: usize) -> u32 {
+    DBUS_VADDR_BASE + (PAGE_SIZE * index) as u32
+}
+
+/// Count flash-mapped pages, de-duplicating mappings which refer to flash page
+/// 0
+#[procmacros::ram]
+pub(super) fn count_effective_flash_pages() -> usize {
+    let mmu_table_ptr = DR_REG_MMU_TABLE as *const u32;
+    let mut page0_seen = false;
+    let mut flash_pages = 0;
+    for i in 0..(TABLE_SIZE - 1) {
+        let mapping = unsafe { mmu_table_ptr.add(i).read_volatile() };
+        if mapping & (ENTRY_INVALID | ENTRY_TYPE) == ENTRY_VALID | ENTRY_ACCESS_FLASH {
+            if mapping & ENTRY_VALID_VAL_MASK == 0 {
+                if page0_seen {
+                    continue;
+                }
+                page0_seen = true;
+            }
+            flash_pages += 1;
+        }
+    }
+    flash_pages
+}
+
+#[procmacros::ram]
+unsafe fn move_flash_to_psram_with_spare(
+    target_entry: usize,
+    psram_page: usize,
+    spare_entry: usize,
+) {
+    let mmu_table_ptr = DR_REG_MMU_TABLE as *mut u32;
+    let target_entry_addr = DBUS_VADDR_BASE + (target_entry * PAGE_SIZE) as u32;
+    let spare_entry_addr = DBUS_VADDR_BASE + (spare_entry * PAGE_SIZE) as u32;
+    unsafe {
+        mmu_table_ptr
+            .add(spare_entry)
+            .write_volatile(psram_page as u32 | ENTRY_ACCESS_SPIRAM);
+        Cache_Invalidate_Addr(spare_entry_addr, PAGE_SIZE as u32);
+        core::ptr::copy_nonoverlapping(
+            target_entry_addr as *const u8,
+            spare_entry_addr as *mut u8,
```

Comment:
`u8` here is quite small. I'm thinking you can get away with `u64`

Reply:
Oh, does this affect how `copy_nonoverlapping` does the copy?

Reply:
Whoops, it's untyped nvm